File: xdialog/src/main/java/com/cl/xdialog/base/XBaseAdapter.java
package com.cl.xdialog.base;

import android.support.annotation.LayoutRes;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.cl.xdialog.XDialog;

import java.util.List;

/**
 * 项目：xDialog
 * 版权：蒲公英公司 版权所有
 * 作者：Arry
 * 版本：1.0
 * 创建日期：2019-08-17
 * 描述：
 * 修订历史：
 */
public abstract class XBaseAdapter<T> extends RecyclerView.Adapter<BindViewHolder> {

    private final int layoutRes;
    private List<T> datas;
    private OnAdapterItemClickListener adapterItemClickListener;
    private XDialog dialog;

    protected abstract void onBind(BindViewHolder holder, int position, T t);

    public XBaseAdapter(@LayoutRes int layoutRes, List<T> datas) {
        this.layoutRes = layoutRes;
        this.datas = datas;
    }

    @Override
    public BindViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new BindViewHolder(LayoutInflater.from(parent.getContext()).inflate(layoutRes, parent, false));
    }

    @Override
    public void onBindViewHolder(final BindViewHolder holder, final int position) {
        onBind(holder, position, datas.get(position));
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                adapterItemClickListener.onItemClick(holder, position, datas.get(position), dialog);
            }
        });
    }

    @Override
    public int getItemCount() {
        return datas.size();
    }

    public void setTDialog(XDialog tDialog) {
        this.dialog = tDialog;
    }

    public interface OnAdapterItemClickListener<T> {
        void onItemClick(BindViewHolder holder, int position, T t, XDialog tDialog);
    }

    public void setOnAdapterItemClickListener(OnAdapterItemClickListener listener) {
        this.adapterItemClickListener = listener;
    }

}

File: xdialog/src/main/java/com/cl/xdialog/base/XBaseAdapter.kt
package com.cl.xdialog.base

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import com.cl.xdialog.R
import com.cl.xdialog.XDialog
import com.cl.xdialog.listener.OnAdapterItemClickListener

/**
 * 项目：xDialog
 * 作者：Arry
 * 版本：1.0
 * 创建日期：2019-08-17
 * 描述：
 * 修订历史：
 */
abstract class XBaseAdapter<T>  @JvmOverloads constructor(
    @param:LayoutRes private val layoutRes: Int,
    private val datas: List<T>,
    @param:LayoutRes private val emptyView: Int? = null
) : RecyclerView.Adapter<BindViewHolder>() {
    private var adapterItemClickListener: OnAdapterItemClickListener<T>? = null
    private var dialog: XDialog? = null


    protected abstract fun onBind(holder: BindViewHolder?, position: Int, t: T)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindViewHolder {
        return if (viewType == VIEW_TYPE_EMPTY) {
            // Return a ViewHolder for the provided empty layout
//            BindViewHolder(emptyView ?: LayoutInflater.from(parent.context).inflate(R.layout.xdialog_layout_empty,
//                parent, false))
//            emptyView
            BindViewHolder(
                LayoutInflater.from(parent.context).inflate(
                    emptyView ?: R.layout.xdialog_layout_empty,
                    parent, false
                )
            )
        } else {
            // Return a ViewHolder for the regular item layout
            BindViewHolder(LayoutInflater.from(parent.context).inflate(layoutRes, parent, false))
        }
    }

    override fun onBindViewHolder(holder: BindViewHolder, position: Int) {
        if (getItemViewType(position) == VIEW_TYPE_ITEM) {
            onBind(holder, position, datas[position])
            holder.itemView.setOnClickListener {
                adapterItemClickListener?.onItemClick(holder, position, datas[position], dialog)
            }
        }
    }

    override fun getItemCount(): Int {
        return if (datas.isEmpty()) 1 else datas.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (datas.isEmpty()) VIEW_TYPE_EMPTY else VIEW_TYPE_ITEM
    }

    val isEmpty: Boolean
        get() = datas.isEmpty()

    fun setTDialog(tDialog: XDialog?) {
        dialog = tDialog
    }

    fun setOnAdapterItemClickListener(listener: OnAdapterItemClickListener<T>?) {
        adapterItemClickListener = listener
    }

    companion object {
        private const val VIEW_TYPE_EMPTY = 0
        private const val VIEW_TYPE_ITEM = 1
    }
}